packages/next/server/web/sandbox/context.ts

```diff
@@ -0,0 +1,240 @@
+import type { Context } from 'vm'
+import { Blob, File, FormData } from 'next/dist/compiled/formdata-node'
+import { readFileSync } from 'fs'
+import { requireDependencies } from './require'
+import { TransformStream } from 'next/dist/compiled/web-streams-polyfill'
+import cookie from 'next/dist/compiled/cookie'
+import * as polyfills from './polyfills'
+import vm from 'vm'
+
+const WEBPACK_HASH_REGEX =
+  /__webpack_require__\.h = function\(\) \{ return "[0-9a-f]+"; \}/g
+
+/**
+ * For a given path a context, this function checks if there is any module
+ * context that contains the path with an older content and, if that's the
+ * case, removes the context from the cache.
+ */
+export function clearModuleContext(path: string, content: Buffer | string) {
+  for (const [key, cache] of caches) {
+    const prev = cache?.paths.get(path)?.replace(WEBPACK_HASH_REGEX, '')
+    if (
+      typeof prev !== 'undefined' &&
+      prev !== content.toString().replace(WEBPACK_HASH_REGEX, '')
+    ) {
+      caches.delete(key)
+    }
+  }
+}
+
+/**
+ * A Map of cached module contexts indexed by the module name. It allows
+ * to have a different cache scoped per module name or depending on the
+ * provided module key on creation.
+ */
+const caches = new Map<
+  string,
+  {
+    context: Context
+    paths: Map<string, string>
+    require: Map<string, any>
+    warnedEvals: Set<string>
+  }
+>()
+
+/**
+ * For a given module name this function will create a context for the
+ * runtime. It returns a function where we can provide a module path and
+ * run in within the context. It may or may not use a cache depending on
+ * the parameters.
+ */
+export function getModuleContext(options: {
+  module: string
+  onWarning: (warn: Error) => void
+  useCache: boolean
+}) {
+  const cache = options.useCache
+    ? caches.get(options.module) || createModuleContext(options)
+    : createModuleContext(options)
+
+  if (options.useCache) {
+    caches.set(options.module, cache)
+  }
```

Comment:
```suggestion
  const cache = options.useCache
    ? caches.get(options.module)
    : createModuleContext(options)

  if (cache === undefined) {
    caches.set(options.module, cache = createModuleContext(options))
  }
```

This calls `caches.set` only when needed